pkg/mqtrigger/messageQueue/nats/nats.go

```diff
@@ -129,112 +129,120 @@ func (nats Nats) Unsubscribe(subscription messageQueue.Subscription) error {
-		// so essentially, function namespace = trigger namespace.
-		url := nats.routerUrl + "/" + strings.TrimPrefix(utils.UrlForFunction(trigger.Spec.FunctionReference.Name, trigger.ObjectMeta.Namespace), "/")
-		nats.logger.Debug("making HTTP request", zap.String("url", url))
-
-		headers := map[string]string{
-			"X-Fission-MQTrigger-Topic":      trigger.Spec.Topic,
-			"X-Fission-MQTrigger-RespTopic":  trigger.Spec.ResponseTopic,
-			"X-Fission-MQTrigger-ErrorTopic": trigger.Spec.ErrorTopic,
-			"Content-Type":                   trigger.Spec.ContentType,
-		}
-
-		// Create request
-		req, err := http.NewRequest("POST", url, bytes.NewReader(msg.Data))
+		cb := func() {
+			// Support other function ref types
+			if trigger.Spec.FunctionReference.Type != fv1.FunctionReferenceTypeFunctionName {
+				nats.logger.Fatal("unsupported function reference type for trigger",
+					zap.Any("function_reference_type", trigger.Spec.FunctionReference.Type),
+					zap.String("trigger", trigger.ObjectMeta.Name))
+			}
 
-		if err != nil {
-			nats.logger.Error("failed to create HTTP request to invoke function",
-				zap.Error(err),
-				zap.String("function_url", url))
-			return
-		}
+			// with the addition of multi-tenancy, the users can create functions in any namespace. however,
+			// the triggers can only be created in the same namespace as the function.
+			// so essentially, function namespace = trigger namespace.
+			url := nats.routerUrl + "/" + strings.TrimPrefix(utils.UrlForFunction(trigger.Spec.FunctionReference.Name, trigger.ObjectMeta.Namespace), "/")
+			nats.logger.Debug("making HTTP request", zap.String("url", url))
+
+			headers := map[string]string{
+				"X-Fission-MQTrigger-Topic":      trigger.Spec.Topic,
+				"X-Fission-MQTrigger-RespTopic":  trigger.Spec.ResponseTopic,
+				"X-Fission-MQTrigger-ErrorTopic": trigger.Spec.ErrorTopic,
+				"Content-Type":                   trigger.Spec.ContentType,
+			}
 
-		for k, v := range headers {
-			req.Header.Set(k, v)
-		}
+			// Create request
+			req, err := http.NewRequest("POST", url, bytes.NewReader(msg.Data))
 
-		var resp *http.Response
-		for attempt := 0; attempt <= trigger.Spec.MaxRetries; attempt++ {
-			// Make the request
-			resp, err = http.DefaultClient.Do(req)
 			if err != nil {
-				nats.logger.Error("sending function invocation request failed",
+				nats.logger.Error("failed to create HTTP request to invoke function",
 					zap.Error(err),
-					zap.String("function_url", url),
-					zap.String("trigger", trigger.ObjectMeta.Name))
-				continue
+					zap.String("function_url", url))
+				return
 			}
-			if resp == nil {
-				continue
+
+			for k, v := range headers {
+				req.Header.Set(k, v)
 			}
-			if err == nil && resp.StatusCode == http.StatusOK {
-				// Success, quit retrying
-				break
+
+			var resp *http.Response
+			for attempt := 0; attempt <= trigger.Spec.MaxRetries; attempt++ {
+				// Make the request
```

Comment:
@harshthakur9030 should this be in a separate PR? I'm actually just moving the entire code block into an anonymous function so that I can conditionally call it as a goroutine, I didn't change any of the existing logic

Reply:
Yes, that bug existed before you touched the code base. Since you have already touched this part of the code base, I was hoping you could do it. Everything remains the same, just the for loop which sets the headers needs to go into the loop which makes the requests.

Reply:
Ok sure, I can do it

Reply:
Done!

Reply:
Thanks !!